### PR TITLE
Avoid materializing all results in Iceberg `$files` system table at once

### PR DESCRIPTION
**Pull request by krvikash**
## Description
Fixes https://github.com/trinodb/trino/issues/15991

The RR is an alternative approach to fix https://github.com/trinodb/trino/issues/15991. another approach https://github.com/trinodb/trino/pull/16063 to fix https://github.com/trinodb/trino/issues/15991

## Release notes

(X) Release notes are required, with the following suggested text:

```markdown
# Iceberg
* Reduce memory usage when reading `$files` system tables. ({issue}`16112`)
```

### CODE REVIEW

**Comment by krvikash** [comment]
CC: @alexjo2144 | @findinpath

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java [hunk 1]
```diff
@@ -15,34 +15,42 @@
 
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.ImmutableMap;
-import io.airlift.slice.Slices;
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:18]: `Avoid materializing all results in Iceberg system table at once` -> `Avoid materializing all table files in Iceberg $files system table at once`

Comment (krvikash) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:18]: somehow GitHub commit did not recognize `$files`. I have updated it now.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java [hunk 2]
```diff
@@ -95,82 +105,185 @@ public ConnectorTableMetadata getTableMetadata()
...
-                pagesBuilder.appendVarbinary(Slices.wrappedBuffer(dataFile.keyMetadata()));
-            }
-            if (checkNonNull(dataFile.splitOffsets(), pagesBuilder)) {
-                pagesBuilder.appendBigintArray(dataFile.splitOffsets());
-            }
-            if (checkNonNull(dataFile.equalityFieldIds(), pagesBuilder)) {
-                pagesBuilder.appendIntegerArray(dataFile.equalityFieldIds());
+        public RecordCursor cursor()
+        {
+            return new IncrementalInMemoryRecordCursor(types, CloseableIterable.withNoopClose(this));
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:144]: why noopclose?
don't we want to close `planFiles` and its `planFilesIterator` when the cursor is being closed?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IncrementalInMemoryRecordCursor.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+public class IncrementalInMemoryRecordCursor
+        implements RecordCursor
+{
+    private final List<Type> types;
+    private final CloseableIterator<List<Object>> records;
+    private List<Object> record;
+    private long completedBytes;
+    private boolean closed;
+
+    public IncrementalInMemoryRecordCursor(List<Type> types, CloseableIterable<List<Object>> records)
```
Comment (alexjo2144) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IncrementalInMemoryRecordCursor.java:44]: Both the `Iterable` and `Iterator` need to be closed

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java [hunk 4]
```diff
@@ -95,82 +105,185 @@ public ConnectorTableMetadata getTableMetadata()
     }
 
     @Override
-    public ConnectorPageSource pageSource(ConnectorTransactionHandle transactionHandle, ConnectorSession session, TupleDomain<Integer> constraint)
+    public RecordCursor cursor(ConnectorTransactionHandle transactionHandle, ConnectorSession session, TupleDomain<Integer> constraint)
     {
+        List<io.trino.spi.type.Type> types = getType(tableMetadata);
+        Map<Integer, Type> idToTypeMapping = getIcebergIdToTypeMapping(icebergTable.schema());
```
Comment (alexjo2144) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:111]: I'd move these down and use `InMemoryRecordSet.cursor()` instead of `PlanFilesIterable` so that you don't need these variables in the empty snapshotId case

```
        if (snapshotId.isEmpty()) {
            return new InMemoryRecordSet(types, ImmutableList.of()).cursor();
        }

        List<io.trino.spi.type.Type> types = getType(tableMetadata);
        Map<Integer, Type> idToTypeMapping = getIcebergIdToTypeMapping(icebergTable.schema());
```

**Review by alexjo2144 — COMMENTED** [review]
This is definitely a better direction that the PageBuilder way

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java (hunk 2 above)
Comment (krvikash) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java]: I was implementing `Iterable` which is why I had to use convert it to `CloseableIterable`.
```java
private static class PlanFilesIterable
            implements Iterable<List<Object>>
```

Now I have modified the code that implements `CloseableIterable`. Thanks for catching this.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java [hunk 5]
```diff
@@ -95,82 +105,185 @@ public ConnectorTableMetadata getTableMetadata()
...
+                public boolean hasNext()
+                {
+                    return !closed && planFilesIterator.hasNext();
+                }
+
+                @Override
+                public List<Object> next()
+                {
+                    DataFile dataFile = planFilesIterator.next().file();
+                    return verifyAndGetColumns(
```
[the remark this answers is not in the document]
Comment (krvikash) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java]: Thanks, @findinpath. This really helps reduce the code complexity. I have modified the code.

**Comment by krvikash** [comment]
Thanks @alexjo2144 | @findinpath for reviewing the PR.

> This is definitely a better direction that the PageBuilder way

I hope we are going forward with this (`RecordCursor`) approach. I will make this as primary PR.

**Comment by krvikash** [comment]
Currently, we materialize all columns from the `$files` table irrespective of what column is asked in the query. Is it worth looking for if we can only materialize the column which is asked?

**Comment by krvikash** [comment]
> Currently, we materialize all columns from the `$files` table irrespective of what column is asked in the query. Is it worth looking for if we can only materialize the column which is asked?

Like, If we query `SELECT file_path from "store_sales$files";` then we collect all columns (My question is here why not just collect the column which is queried for, `file_path` in this case) from $files table and pass it to the caller. and the caller will display according to what field we have in the SELECT query.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IncrementalInMemoryRecordCursor.java [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+    @Override
+    public boolean isNull(int field)
+    {
+        checkArgument(record != null, "no current record");
+        return record.get(field) == null;
+    }
+
+    @Override
+    public void close()
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IncrementalInMemoryRecordCursor.java:142]: except `close()` are there major differences between this class and the methods from `io.trino.spi.connector.InMemoryRecordSet` ? Would it make sense to rather extend the before mentioned class instead writing a new one?

Comment (krvikash) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IncrementalInMemoryRecordCursor.java]: Yes, only the `close()` method was additional. I have renamed it with `CloseableInMemoryRecordCursor` which will extend the `InMemoryRecordSet.InMemoryRecordCursor` and implement the `close()` method.

Comment (krvikash) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IncrementalInMemoryRecordCursor.java]: Thanks for pointing it out. The code is much reduced now.

**Comment by krvikash** [comment]
Addressed comments and rebased with master.

File: core/trino-spi/src/main/java/io/trino/spi/connector/InMemoryRecordSet.java [hunk 7]
```diff
@@ -68,15 +68,15 @@ public RecordCursor cursor()
         return new InMemoryRecordCursor(types, records.iterator());
     }
 
-    private static class InMemoryRecordCursor
+    public static class InMemoryRecordCursor
```
Comment (homar) [inline core/trino-spi/src/main/java/io/trino/spi/connector/InMemoryRecordSet.java:71]: I think that changes in SPI should be in a separate commit

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java [hunk 8]
```diff
@@ -95,82 +106,153 @@ public ConnectorTableMetadata getTableMetadata()
...
+                {
+                    PlanFilesIterable.super.close();
+                    planFiles = null;
+                    planFilesIterator = null;
+                    closed = true;
+                }
+            };
+        }
+
+        private List<Object> getRow(DataFile dataFile)
```
Comment (homar) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:184]: `mapToRow` or `toRow` ?

**Review by homar — APPROVED** [review]

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CloseableInMemoryRecordCursor.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    {
+        super(types, records.iterator());
+    }
+
+    @Override
+    public void close()
+    {
+        try {
+            if (!closed) {
+                ((CloseableIterator<?>) records).close();
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CloseableInMemoryRecordCursor.java:39]: If we'd extend from

```
public class CloseableInMemoryRecordCursor
        extends InMemoryRecordSet.InMemoryRecordCursor<CloseableIterator<? extends List<?>>>
```

```
InMemoryRecordCursor<T extends  Iterator<? extends List<?>>>
...
        protected T records;

```

then this cast would not be necessary anymore.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CloseableInMemoryRecordCursor.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.List;
+
+public class CloseableInMemoryRecordCursor
+        extends InMemoryRecordSet.InMemoryRecordCursor
+{
+    private boolean closed;
+
+    public CloseableInMemoryRecordCursor(List<Type> types, Iterable<? extends List<?>> records)
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CloseableInMemoryRecordCursor.java:29]: Why not provide a `CloseableIterator<? extends List<?>>` instead of providing the whole `Iterable<? extends List<?>>` ?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CloseableInMemoryRecordCursor.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+import io.trino.spi.connector.InMemoryRecordSet;
+import io.trino.spi.type.Type;
+import org.apache.iceberg.io.CloseableIterator;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.List;
+
+public class CloseableInMemoryRecordCursor
+        extends InMemoryRecordSet.InMemoryRecordCursor
```
Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CloseableInMemoryRecordCursor.java:25]: The class `InMemoryRecordCursor` is a `RecordCursor` which is `Closeable` .
We should find a better name.

optional : `CloseableIterableInMemoryRecordCursor`

**Comment by krvikash** [comment]
Hi @ebyhr, When you get time, could you please take a look at this PR?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java [hunk 12]
```diff
@@ -95,82 +106,153 @@ public ConnectorTableMetadata getTableMetadata()
...
+        }
+
+        private Block getIntegerBigintMapBlock(Map<Integer, Long> value)
+        {
+            if (value == null) {
+                return null;
             }
-            if (checkNonNull(dataFile.keyMetadata(), pagesBuilder)) {
-                pagesBuilder.appendVarbinary(Slices.wrappedBuffer(dataFile.keyMetadata()));
+            else {
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:210]: nit: Remove redundant `else`.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java [hunk 13]
```diff
@@ -95,82 +106,153 @@ public ConnectorTableMetadata getTableMetadata()
...
+        }
+
+        private Block getIntegerVarcharMapBlock(Map<Integer, ByteBuffer> value)
+        {
+            if (value == null) {
+                return null;
             }
-            if (checkNonNull(dataFile.equalityFieldIds(), pagesBuilder)) {
-                pagesBuilder.appendIntegerArray(dataFile.equalityFieldIds());
+            else {
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:220]: nit: Remove redundant `else`

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java [hunk 14]
```diff
@@ -95,82 +106,153 @@ public ConnectorTableMetadata getTableMetadata()
...
+                                                idToTypeMapping.get(entry.getKey()), Conversions.fromByteBuffer(idToTypeMapping.get(entry.getKey()), entry.getValue())))),
+                        typeOperators);
             }
-            pagesBuilder.endRow();
-        });
+        }
 
-        return pagesBuilder.build();
-    }
+        private static Block toIntegerBigintMapBlock(Map<Integer, Long> values, TypeOperators typeOperators)
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:231]: Can we avoid taking `TypeOperators` argument by removing `static`?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java [hunk 15]
```diff
@@ -95,82 +106,153 @@ public ConnectorTableMetadata getTableMetadata()
...
+            blockBuilder.closeEntry();
+            return mapType.getObject(blockBuilder, 0);
+        }
 
-    private static boolean checkNonNull(Object object, PageListBuilder pagesBuilder)
-    {
-        if (object == null) {
-            pagesBuilder.appendNull();
-            return false;
+        private static Block toIntegerVarcharMapBlock(Map<Integer, String> values, TypeOperators typeOperators)
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:244]: Can we avoid taking `TypeOperators` argument by removing `static`?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java [hunk 16]
```diff
@@ -95,82 +106,153 @@ public ConnectorTableMetadata getTableMetadata()
...
+        {
+            if (value == null) {
+                return null;
             }
-            if (checkNonNull(dataFile.equalityFieldIds(), pagesBuilder)) {
-                pagesBuilder.appendIntegerArray(dataFile.equalityFieldIds());
+            else {
+                return toIntegerVarcharMapBlock(
+                        value.entrySet().stream()
+                                .filter(entry -> idToTypeMapping.containsKey(entry.getKey())).collect(toImmutableMap(
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:223]: nit: Wrap before `.collect`.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java [hunk 17]
```diff
@@ -95,82 +106,153 @@ public ConnectorTableMetadata getTableMetadata()
...
+            columns.add(getIntegerBigintMapBlock(dataFile.columnSizes()));
+            columns.add(getIntegerBigintMapBlock(dataFile.valueCounts()));
+            columns.add(getIntegerBigintMapBlock(dataFile.nullValueCounts()));
+            columns.add(getIntegerBigintMapBlock(dataFile.nanValueCounts()));
+            columns.add(getIntegerVarcharMapBlock(dataFile.lowerBounds()));
+            columns.add(getIntegerVarcharMapBlock(dataFile.upperBounds()));
+            columns.add(dataFile.keyMetadata());
+            columns.add(dataFile.splitOffsets());
+            columns.add(dataFile.equalityFieldIds());
+            checkArgument(columns.size() == types.size(), "Expected %s values in row, but got %s values", types.size(), columns.size());
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:201]:
```suggestion
            checkArgument(columns.size() == types.size(), "Expected %s types in row, but got %s values", types.size(), columns.size());
```

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java [hunk 18]
```diff
@@ -95,82 +106,153 @@ public ConnectorTableMetadata getTableMetadata()
...
+        private final List<io.trino.spi.type.Type> types;
+        private final TypeOperators typeOperators;
+        private final Map<Integer, Type> idToTypeMapping;
+        private CloseableIterator<FileScanTask> planFilesIterator;
+        private boolean closed;
+
+        public PlanFilesIterable(CloseableIterable<FileScanTask> planFiles, Map<Integer, Type> idToTypeMapping, List<io.trino.spi.type.Type> types, TypeOperators typeOperators)
+        {
+            this.planFiles = requireNonNull(planFiles, "planFiles is null");
+            this.types = requireNonNull(types, "types is null");
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:141]: Add `ImmutableList.copyOf`.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java [hunk 19]
```diff
@@ -95,82 +106,153 @@ public ConnectorTableMetadata getTableMetadata()
...
+            this.typeOperators = requireNonNull(typeOperators, "typeOperators is null");
+            this.idToTypeMapping = requireNonNull(idToTypeMapping, "idToTypeMapping is null");
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:143]: Add `ImmutableMap.copyOf`.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java [hunk 20]
```diff
@@ -95,82 +106,153 @@ public ConnectorTableMetadata getTableMetadata()
...
+            this.planFilesIterator = planFiles.iterator();
```
Comment (krvikash) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:144]: `planFilesIterator` should be initialize in `iterator()`

**Review by ebyhr — COMMENTED** [review]
Just skimmed.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java [hunk 21]
```diff
@@ -95,82 +106,153 @@ public ConnectorTableMetadata getTableMetadata()
...
+        }
+
+        public RecordCursor cursor()
+        {
+            return new CloseableInMemoryRecordCursor(types, this);
```
[the remark this answers is not in the document]
Comment (krvikash) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:149]: `PlanFilesIterable.super.close();` will close `planFiles` and `planFilesIterator`. Is this what you are referring to?

[the remark this answers is not in the document]
Comment (findinpath) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:149]:
```
public RecordCursor cursor()
        {
            return new InMemoryRecordSet.InMemoryRecordCursor<CloseableIterator<? extends List<?>>>(types, this.iterator()){

                @Override
                public void close()
                {
                    try {
                        records.close();
                    }
                    catch (IOException e) {
                        throw new UncheckedIOException("Failed to close cursor", e);
                    }
                }
            };
        }
```

I hope this makes sense

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java (hunk 8 above)
Comment (krvikash) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:184]: how about `getRecord`? Since we are using `RecordCursor` to iterate over.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java (hunk 18 above)
Comment (krvikash) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java]: Hi @ebyhr, What was the reasoning behind making a copy of `types` and `idToTypeMapping`?

**Comment by krvikash** [comment]
Thanks, @ebyhr | @findinpath | @homar for the reviews. I have addressed the comments.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CloseableInMemoryRecordCursor.java (hunk 11 above)
Comment (krvikash) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/CloseableInMemoryRecordCursor.java]: Not needed after latest chage.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java [hunk 22]
```diff
@@ -95,82 +107,157 @@ public ConnectorTableMetadata getTableMetadata()
...
+                                    Map.Entry<Integer, ByteBuffer>::getKey,
+                                    entry -> Transforms.identity().toHumanString(
+                                            idToTypeMapping.get(entry.getKey()), Conversions.fromByteBuffer(idToTypeMapping.get(entry.getKey()), entry.getValue())))));
+        }
 
-        return pagesBuilder.build();
-    }
+        private Block toIntegerBigintMapBlock(Map<Integer, Long> values)
+        {
+            MapType mapType = new MapType(INTEGER, VARCHAR, typeOperators);
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:238]: This creates a new `MapType` object every time. Any reason avoiding `TypeManager#getType`? Same for `toIntegerVarcharMapBlock` method.

Comment (findinpath) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:238]: Did you mean `new MapType(INTEGER, BIGINT, typeOperators);` ?

Comment (krvikash) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:238]: Yes, I am making the change and addressing it as @ebyhr suggested.

Comment (krvikash) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java]: I was not aware of `TypeManager#getType`. Changed the code to use `TypeManager#getType`.

File: core/trino-spi/src/main/java/io/trino/spi/connector/InMemoryRecordSet.java [hunk 23]
```diff
@@ -68,15 +68,15 @@ public RecordCursor cursor()
         return new InMemoryRecordCursor(types, records.iterator());
     }
 
-    private static class InMemoryRecordCursor
+    public static class InMemoryRecordCursor<T extends Iterator<? extends List<?>>>
             implements RecordCursor
     {
         private final List<Type> types;
-        private final Iterator<? extends List<?>> records;
+        protected T records;
```
Comment (findinpath) [inline core/trino-spi/src/main/java/io/trino/spi/connector/InMemoryRecordSet.java:75]: add `final`

**Review by findinpath — APPROVED** [review]

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java (hunk 18 above)
Comment (ebyhr) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java]: This is the convention to prevent unexpected modification by other methods.

**Review by ebyhr — APPROVED** [review]

File: core/trino-spi/src/main/java/io/trino/spi/connector/InMemoryRecordSet.java [hunk 24]
```diff
@@ -68,15 +68,15 @@ public RecordCursor cursor()
         return new InMemoryRecordCursor(types, records.iterator());
     }
 
-    private static class InMemoryRecordCursor
+    public static class InMemoryRecordCursor<T extends Iterator<? extends List<?>>>
```
Comment (findepi) [inline core/trino-spi/src/main/java/io/trino/spi/connector/InMemoryRecordSet.java:71]: The class became generic, but so there are now warnings wherever it's used as a raw class (eg 3 lines above).

BTW it seems the generics are not leveraged, since T shows up only in the constructor declaration.
https://github.com/trinodb/trino/pull/16295
